### ISSUE TITLE
Bootstrap 3 cascading menus

### ISSUE THREAD

**Issue by HarryR7** [web2py/web2py#938]
Since web2py's upgrade to bootstrap 3, the cascading menus became invisible (submenu not implemented in bootstrap 3). I found the following tweak 
< style>
      .dropdown-submenu{position:relative;}
.dropdown-submenu>.dropdown-menu{top:0;left:100%;margin-top:-6px;margin-left:-1px;-webkit-border-radius:0 6px 6px 6px;-moz-border-radius:0 6px 6px 6px;border-radius:0 6px 6px 6px;}
.dropdown-submenu:hover>.dropdown-menu{display:block;}
.dropdown-submenu>a:after{display:block;content:" ";float:right;width:0;height:0;border-color:transparent;border-style:solid;border-width:5px 0 5px 5px;border-left-color:#cccccc;margin-top:5px;margin-right:-10px;}
.dropdown-submenu:hover>a:after{border-left-color:#ffffff;}
.dropdown-submenu.pull-left{float:none;}.dropdown-submenu.pull-left>.dropdown-menu{left:-100%;margin-left:10px;-webkit-border-radius:6px 0 6px 6px;-moz-border-radius:6px 0 6px 6px;border-radius:6px 0 6px 6px;}
< /style>
to be implemented in web2py.css or layout.html i order to fix the issue.
HTH

**Comment by niphlod** [comment]
won't be fixed. bs3 doesn't support 3rd level menus.

**Comment by HarryR7** [comment]
Try the tweak!

**Comment by niphlod** [comment]
not interested unless you're willing to support nested multilevel menus on bs3 at each release. In that case, you should submit your tweaks to the bs3 repository and not here anyway....

**Comment by HarryR7** [comment]
Not interested either. I was happy to contribute...
> If you benefit from it, I hope you will feel encouraged to pay it forward by contributing back to society in whatever form you choose.

**Comment by BuhtigithuB** [comment]
Since Niphold is reluctant to include it, you may consider creating a slice
about this fix with some screenshot...

Richard

**Comment by leonelcamara** [comment]
This is a bad idea for yet another reason. It depends on hover which just doesn't work on mobile. The menu style the old welcome application had is outdated and no point in insisting on keeping it that way.

**Comment by niphlod** [comment]
re @everybody : it's not niphlod being nitpicky, it's niphlod making some points:
- we're in the business of creating a web framework
- we're not in the business of creating YOUR app
- we're not in the business of css framework, we just ADOPT one
- as other pieces involved, it's best to let everyone what it does best. We're not asking boostrap developers to make life easier for pythonistas. They're not in the busines of creating a web framework.

That's the beauty of opensource. You can contribute everywhere and if there are enough "experts" on the developer-list it and your fix is good, it gets into the code, gets maintained, tested, it eventually gains traction, gets wonderful, etc etc etc. 

tl;dr: let bootstrap developer know about your fix: if it's smart, it'll probably get included and web2py will adopt it. There's absolutely no reason to ask web2py to maintain a fork of bootstrap. Contribute there, if you're happy to contribute.

**Comment by BuhtigithuB** [comment]
You are right Niphold and it is good idea to remember OOS HOW-TO
contribute...

Let get back Super Fish (just kidding!!)

Richard

**Comment by HarryR7** [comment]
My problem is that nested functionality of MENU (I was counting on it for desktop browsers) was broken after switch to bootstrap3. I like bootstrap3 but I do not consider submenus an outdated feature, and since is mentioned in the book, it should be functionally available. 
Or 
> "It's not a bug, it's a feature..." ;-)

disclaimer in the next release of the book, please!
No need to include the fix anywhere, no need for a bootstrap fork, for now.

**Comment by HarryR7** [comment]
Be gentle with me, I'm new in this GitHub thing. It seems I have to create a fork in order to really contribute code...

**Comment by niphlod** [comment]
let's get picky for valid reasons please. 
"nested functionality of MENU was broken": not really. 
MENU generates a tree-like of ``<ul><li>``. 
there's nothing in the book sustaining that a 10fold nested structure of menu won't be generated, and in fact, you can still generate it. How to hook it up with your personal favourite css or js menu is up to your app.

totally different argument: bs3 doesn't support past the 2nd level. if 3rd level is important, by all means, ditch bootstrap3: fortunately there are tons of alternatives.

welcome IS NOT your app. welcome is a scaffolding app. *Scaffolding* synonims: frame, scheme, structure, scenery, and... **STAGE**. 

my personal favourite (I'll repeat this to every user asking for an out-of-scope feature).....

You are a director in the need of hosting a show for the public, with a budget.
You need to choose a theater. web2py is one. it's free but that's beside the point.
Is the **theater** fine ? 
stick with web2py
Don't like the theater ? 
go with another web framework.

Enter the theater....... you see the stage. The stage was recently renovated. BS3 stage was requested by lots of "directors" like you. It has because who crafted the stage does it really well, AND sells it for free (and ships it, and mounts it, again, FOR FREE). Other directors sold zillions of tickets with little or no time involved in stage-painting: they got more budget to spend on casting and screenplay. Directors asking for dragons spitting fire were told that the stage floor was wood AND it was a security threat because the average Joe going to a shows isn't a firefighter.

a) Theater is fine but stage needs **refinement** ? You prefer purple clouds instead of white ones on the index ? hurray, *scaffolding* saved your time and budget.
b) Theater is fine, but stage needs [Drogon](http://gameofthrones.wikia.com/wiki/Drogon) AND [Smaug](http://en.wikipedia.org/wiki/Smaug) spitting fire from above, a helium baloon on the middle, a [Weirwood](http://gameofthrones.wikia.com/wiki/Weirwood) on both sides,  because your show is open to firefighters only? Need also an occasional appearance of [manticores](http://gameofthrones.wikia.com/wiki/Manticores) pinching actors at random times to spice things up? ask another stage-builder or DIY.

The **theater** isn't getting in the way, but you, as a director, will need to work a taddle bit more.

**Comment by HarryR7** [comment]
You are right!  Nicely put!
